crates/ruff_linter/src/importer/mod.rs

```diff
@@ -135,71 +135,76 @@ impl<'a> Importer<'a> {
             self.stylist,
         )?;
 
-        // Import the `TYPE_CHECKING` symbol from the typing module.
-        let (type_checking_edit, type_checking) =
-            if let Some(type_checking) = Self::find_type_checking(at, semantic)? {
-                // Special-case: if the `TYPE_CHECKING` symbol is imported as part of the same
-                // statement that we're modifying, avoid adding a no-op edit. For example, here,
-                // the `TYPE_CHECKING` no-op edit would overlap with the edit to remove `Final`
-                // from the import:
-                // ```python
-                // from __future__ import annotations
-                //
-                // from typing import Final, TYPE_CHECKING
-                //
-                // Const: Final[dict] = {}
-                // ```
-                let edit = if type_checking.statement(semantic) == import.statement {
-                    None
+        // Add the import to a `TYPE_CHECKING` block.
+        if let Some(block) = self.preceding_type_checking_block(at) {
+            // Add the import to the existing `TYPE_CHECKING` block.
+            Ok(TypingImportEdit {
+                type_checking_edit: None,
+                add_import_edit: self.add_to_type_checking_block(&content, block.start()),
+            })
+        } else {
+            // Import the `TYPE_CHECKING` symbol from the typing module.
+            // TODO: Should we provide an option to avoid this import?
+            //       E.g. either through an explicit setting, or implicitly
+            //       when `typing` isn't part of the exempt modules and there
+            //       are no other existing runtime imports of `typing`.
+            let (type_checking_edit, type_checking) =
+                if let Some(type_checking) = Self::find_type_checking(at, semantic)? {
+                    // Special-case: if the `TYPE_CHECKING` symbol is imported as part of the same
+                    // statement that we're modifying, avoid adding a no-op edit. For example, here,
+                    // the `TYPE_CHECKING` no-op edit would overlap with the edit to remove `Final`
+                    // from the import:
+                    // ```python
+                    // from __future__ import annotations
+                    //
+                    // from typing import Final, TYPE_CHECKING
+                    //
+                    // Const: Final[dict] = {}
+                    // ```
+                    let edit = if type_checking.statement(semantic) == import.statement {
+                        None
+                    } else {
+                        Some(Edit::range_replacement(
+                            self.locator.slice(type_checking.range()).to_string(),
+                            type_checking.range(),
+                        ))
```

Comment:
By restructuring the logic, this no-op edit would now only happen if we don't have an existing type checking block. I'm not sure whether this is problematic or why we're adding a no-op edit anyways (maybe to ensure no other edit is allowed to remove the import?).

If this no-op edit indeed accomplishes something useful, we may need a more sophisticated change, which allows `find_type_checking` to return a name based on an assignment target in addition to an import. This would allow us to write separate fix logic for when it's imported and when it's defined in the same file.

Reply:
The other thing that bothered me, is that this hack does not seem very reliable, since the type checking block we're moving the imports to might not correspond to the `typing`/`TYPE_CHECKING` import we find in the first step...

Maybe we revert it back to my earlier refactored version, but when we have an existing type checking block we lookup the relevant statement from the binding and emit a no-op edit for that specific statement if it doesn't overlap with one of the import edits. That seems a lot more robust to me.

Reply:
I had another little think about this. I don't think we have to worry about the no-op edit when we find a preceding type checking block. If another edit would remove the binding we're referencing, it would necessarily also have to remove the type checking block itself (otherwise there would still be a reference to the binding), but then the edit that removes the type checking block would overlap with our edit to add new import to it, so we don't need to artificially create another overlap.

In the case where there isn't a type checking block, but there is a unused `TYPE_CHECKING` import, it makes sense, since we don't want to let F401 remove the import when our edit adds a new reference to it.

Reply:
Yeah, I believe this is to prevent things like:

```python
from typing import TYPE_CHECKING

from foo import Bar

def func() -> Bar: ...
```

If we don't add the no-op edit, `TYPE_CHECKING` might be removed, leaving us with:

```python
if TYPE_CHECKING:
  from foo import Bar

def func() -> "Bar": ...
```